src/transactions/TransactionFrame.cpp

```diff
@@ -684,6 +692,12 @@ TransactionFrame::apply(Application& app, AbstractLedgerTxn& ltx,
         auto signaturesValid = cv >= (ValidationType::kInvalidPostAuth) &&
                                processSignatures(signatureChecker, ltxTx);
 
+        if (cv >= ValidationType::kInvalidRemoveOneTimeSigners &&
+            ltxTx.loadHeader().current().ledgerVersion >= 13)
+        {
+            removeOneTimeSignerFromAllSourceAccounts(ltxTx);
```

Comment:
Also, we do not need the condition `cv >= ValidationType::kInvalidRemoveOneTimeSigners` (just that protocol 13 always removes one time signers):

when transactions get applied, we already charged the fee (in a different step), and we have to guarantee that the transaction will never get accepted in a future ledger, so we can take advantage of that.

It's a bit strange/unfortunate (I know the code is ultra confusing), `commonValidPreSeqNum` today can fail during consensus even though the only way it should be able to fail should be `txNO_ACCOUNT` (issue for a different time).
But even in the case when the source account on the transaction was removed, it doesn't mean we should not remove the one time signer on the other source accounts.

@jonjove I think we missed this

Reply:
@MonsieurNicolas Just so I understand this - Is the reason `commonValidPreSeqNum` will not fail during consensus with the errors other than `txNO_ACCOUNT` because those transactions would not have been accepted in the transaction set?

Reply:
no: they got accepted but the failure is irrelevant because we should not be able to replay the transaction anyways @sisuresh